torch/csrc/autograd/record_function_ops.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+#include <torch/csrc/autograd/record_function.h>
+#include <torch/csrc/utils/future.h>
+
+namespace torch {
+namespace autograd {
+namespace profiler {
+// Creates a new profiling scope using RecordFunction and invokes its starting
+// callbacks.
+TORCH_API at::Tensor record_function_enter(const std::string& name);
+
+// Cast Tensor that was created with at::cpp_custom_type_hack back to
+// RecordFunction. This is a temporary workaround until RecordFunction is
+// registered as a custom C++ class.
+TORCH_API RecordFunction& getRecordFunctionFromTensor(const at::Tensor& handle);
+
+// Schedules RecordFunction's end callbacks to be run on completion of a future.
+template <typename T>
+TORCH_API void call_end_callbacks_on_fut(
+    const at::Tensor& handle,
+    const std::shared_ptr<torch::utils::Future<T>> fut) {
+  // Add a callback onto the future to mark run RecordFunction's end callbacks
+  // when the future is completed.
+  fut->addCallback(
+      // Copy handle by value to persist after the python context manager is
+      // exited.
+      [handle](
+          const T& /* unused */,
+          const c10::optional<torch::utils::FutureError>& /* unused */) {
+        TORCH_INTERNAL_ASSERT(
+            handle.defined(),
+            "Undefined RecordFunction handle. This can happen if the handle is "
+            "not correctly persisted and is destroyed before the future is "
+            "realized.");
```

Comment:
By above do you mean before we call `fut->addCallback()`? I placed it inside the callback code because I want to ensure `handle` is defined when the callback actually runs. 

I was running into the error where the python `RecordFunction` would go out of scope, and since I had copied `handle` by reference, it would be invalid when the callback runs at a later time (i.e. after the python `with record_function...` scope is exited. So copying by value and adding the assert here could help detect this error again if the code is modified in the future.